ttnn/cpp/ttnn/operations/moreh/moreh_linear_backward/moreh_linear_backward.cpp

```diff
@@ -0,0 +1,175 @@
+            tt::operations::primary::is_same_shape(weight, weight_grad_tensor),
+            "both tensors should be the same shape");
+    }
+
+    if (bias_grad.has_value()) {
+        const auto& bias_grad_tensor = bias_grad.value();
+        TT_FATAL(
+            tt::operations::primary::is_scalar(bias_grad_tensor) ||
+                tt::operations::primary::is_1d_tensor(bias_grad_tensor),
+            "bias_grad tensor should be 1d or scalar");
+    }
+}
+
+std::vector<int64_t> find_reduce_dim(
+    const tt::tt_metal::LegacyShape& a_shape, const tt::tt_metal::LegacyShape& b_shape) {
+    std::vector<uint32_t> a_dim(tt::tt_metal::MAX_NUM_DIMENSIONS, 1);
+    std::vector<uint32_t> b_dim(tt::tt_metal::MAX_NUM_DIMENSIONS, 1);
+    get_tensor_dim(a_dim, a_shape);
+    get_tensor_dim(b_dim, b_shape);
+    int32_t rank = std::max(a_shape.rank(), b_shape.rank());
+    log_debug(tt::LogOp, "find_reduce_dim :{} rank {} a {} b {}", __LINE__, rank, a_shape.rank(), b_shape.rank());
+    std::vector<int64_t> dims;
+    // batch dims
+    for (int i = 0; i < rank - 2; ++i) {
+        int idx = rank - 1 - i;
+        TT_ASSERT(idx >= 0);
+        if (a_dim[idx] != b_dim[idx]) {
+            dims.push_back(i);
+            log_debug(tt::LogOp, "find_reduce_dim :{} push {} dim", __LINE__, i);
+        }
+    }
+    return dims;
+}
+
+bool is_same_batch_dim(const Tensor& tensor_a, const Tensor& tensor_b) {
+    // check batch dims
+    const auto& a_shape = tensor_a.get_shape().value;
+    const auto& b_shape = tensor_b.get_shape().value;
+    std::vector<uint32_t> a_dim(tt::tt_metal::MAX_NUM_DIMENSIONS, 1);
+    std::vector<uint32_t> b_dim(tt::tt_metal::MAX_NUM_DIMENSIONS, 1);
+    get_tensor_dim(a_dim, a_shape);
+    get_tensor_dim(b_dim, b_shape);
+    for (auto i = 2; i < tt::tt_metal::MAX_NUM_DIMENSIONS; ++i) {
+        if (a_dim[i] != b_dim[i]) {
+            log_debug(tt::LogOp, "{}:{} {} a_dim {} - b_dim {}", __func__, __LINE__, i, a_dim[i], b_dim[i]);
+            return false;
+        }
+    }
+    log_debug(tt::LogOp, "{}:{} batch dims are the same.", __func__, __LINE__);
+    return true;
+}
+
+std::vector<std::optional<Tensor>> MorehLinearBackward::invoke(
+    const Tensor& output_grad,
+    const Tensor& input,
+    const Tensor& weight,
+    const std::vector<bool>& are_required_outputs,
+    const std::optional<Tensor>& bias,
+    const std::optional<Tensor>& input_grad,
+    const std::optional<Tensor>& weight_grad,
+    const std::optional<Tensor>& bias_grad,
+    const std::optional<ttnn::MemoryConfig>& input_grad_mem_config,
+    const std::optional<ttnn::MemoryConfig>& weight_grad_mem_config,
+    const std::optional<ttnn::MemoryConfig>& bias_grad_mem_config,
+    const DeviceComputeKernelConfig compute_kernel_config) {
+    std::vector<std::optional<Tensor>> result(3);
+    const auto [input_required_grad, weight_required_grad, bias_required_grad] =
+        get_required_outputs(are_required_outputs);
+
+    TT_ASSERT(
```

Comment:
```suggestion
    TT_FATAL(
```

Reply:
resolved in [b87ffe7](https://github.com/tenstorrent/tt-metal/pull/12812/commits/b87ffe749acd90b038cdc391f7cb2e0b8e81f3cb)